src/main/java/io/r2dbc/postgresql/PostgresqlConnectionFactoryProvider.java

```diff
@@ -48,6 +55,16 @@
      */
     public static final Option<String> SCHEMA = Option.valueOf("schema");
 
+    /**
+     * Connection options which are applied once after the connection has been created.
+     * Format is '{key}={value};...;{key}={value}'
+     */
+    public static final Option<String> OPTIONS = Option.valueOf("options");
+
+    private static final Pattern itemPattern = Pattern.compile(";");
```

Comment:
I'm not exactly happy about nesting `options` and assuming a nested, encoded string as the resulting format would be:

```
r2dbc:postgres://…/database?options=lock_timeout%3D5s%3Bstatement_timeout%3D60000
```

After having a deeper look into the docs I totally understand why it's not feasible to include a pre-compiled property list.

How about not adding `options` to the URL and leaving options configuration for programmatic config only?

Reply:
Indeed, such a URL would be pretty nasty. 

But my primary vision was not to put the options into the URL, but rather use `R2dbcProperties.properties` and a separate entry the application.yml:

```
spring.r2dbc.properties.options: "lock_timeout=1s;statement_timeout=30min"
```

into my application.yml.

The ultimate goal would be to go one step further and remove the "options" nesting altogether:

```
spring.r2dbc.properties.lock_timeout: "1min"
spring.r2dbc.properties.statement_timeout: "30min"
```

This – then – could be also achieved using query parameters in the R2DBC URL.

But it would require some changes in the behaviour of `PostgresqlConnectionFactoryProvider`. Currently, the code only reads a select set of supported fields like `CONNECT_TIMEOUT` from the `ConnectionFactoryOptions`, so putting something into `spring.r2dbc.properties` or query parameters in `spring.r2dbc.url` is pointless ATM. Supporting them would require `PostgresqlConnectionFactoryProvider` to scan and use all options in  `ConnectionFactoryOptions` (and the class currently does not even allow option iteration as the options map is not exposed). 

If you like, I can move the options handling into this direction.

Reply:
Spring (Boot/Data) and the Postgres driver are two independent pieces. Let's try to address the configuration issue in a programmatic-way only by moving the problem into the config abstractions of Spring Boot by e.g. adding there some sort of `ConnectionFactoryOptionsCustomizer`.

In the driver, we would expect `options`  to be `Map<String, String>` (well, we cannot check for generics, but that's a different issue) instead of introducing a nested key-value format to the URL. The options constant would be defined as `public static final Option<Map<String, String>> OPTIONS = Option.valueOf("options");` and `PostgresqlConnectionFactoryProvider ` could reject non-`Map` values.

By doing so, we simplify the problem to some extent and allow augmentation of `ConnectionFactoryOptions` on a different level.